src/drivers/src/digest.rs

```diff
@@ -0,0 +1,140 @@
+                                return Err(SyscallError::ResourceBusy);
+                            }
+                            self.current_user.set(Some(caller_id));
+
+                            let digest_mode = match r2 {
+                                0 => DigestMode::Sha1,
+                                1 => DigestMode::Sha256,
+                                _ => return Err(SyscallError::InvalidArgument),
+                            };
+
+                            try!(self.engine.borrow_mut().initialize(digest_mode));
+
+                            Ok(0)
+                        })
+                        .unwrap_or(Err(SyscallError::Unknown))
+                }
+                // Feed data from input buffer (arg: number of bytes)
+                1 => {
+                    self.apps
+                        .enter(caller_id, |app_data, _| {
+                            if self.current_user.get() != Some(caller_id) {
+                                return Err(SyscallError::InvalidState);
+                            }
+
+                            let app_data: &mut AppData = app_data;
+
+                            let input_buffer = match app_data.input_buffer {
+                                Some(ref slice) => slice,
+                                None => return Err(SyscallError::InvalidState),
+                            };
+
+                            let input_len = r2;
+                            if input_len > input_buffer.len() {
+                                return Err(SyscallError::OutOfRange);
+                            }
+
+                            try!(self.engine
+                                .borrow_mut()
+                                .update(&input_buffer.as_ref()[..input_len]));
+
+                            Ok(0)
+                        })
+                        .unwrap_or(Err(SyscallError::Unknown))
+                }
+                // Finalize hash and output to output buffer (arg: unused)
+                2 => {
+                    self.apps
+                        .enter(caller_id, |app_data, _| {
+                            if self.current_user.get() != Some(caller_id) {
+                                return Err(SyscallError::InvalidState);
+                            }
+
+                            let app_data: &mut AppData = app_data;
+
+                            let output_buffer = match app_data.output_buffer {
+                                Some(ref mut slice) => slice,
+                                None => return Err(SyscallError::InvalidState),
+                            };
+
+                            try!(self.engine.borrow_mut().finalize(output_buffer.as_mut()));
+
+                            Ok(0)
+                        })
+                        .unwrap_or(Err(SyscallError::Unknown))
+                }
+                _ => Err(SyscallError::NotImplemented),
+            }
```

Comment:
I'm not sure why rustfmt things it should double-indent the match body, or why it should leave this closing bracket here hanging.